src/server-core/src/lib.rs

```diff
@@ -156,52 +157,104 @@ pub async fn listen(
+                //
+                // If set_nodelay fails, it's a programming error, so panic.
+                //
+                // [0]: https://news.ycombinator.com/item?id=10608356
+                conn.set_nodelay(true).expect("set_nodelay failed");
+                // Enable TCP keepalives to avoid any idle connection timeouts that may
+                // be enforced by networking devices between us and the client. Idle SQL
+                // connections are expected--e.g., a `SUBSCRIBE` to a view containing
+                // critical alerts will ideally be producing no data most of the time.
+                if let Err(e) = SockRef::from(&conn).set_tcp_keepalive(&KEEPALIVE) {
+                    error!("failed enabling keepalive: {e}");
+                    continue;
+                }
+                let fut = server.handle_connection(conn);
+                set.spawn_named(|| &task_name, async {
+                    if let Err(e) = fut.await {
+                        debug!(
+                            "error handling connection in {}: {}",
+                            S::NAME,
+                            e.display_with_causes()
+                        );
+                    }
+                });
             }
-        };
-        // Set TCP_NODELAY to disable tinygram prevention (Nagle's
-        // algorithm), which forces a 40ms delay between each query
-        // on linux. According to John Nagle [0], the true problem
-        // is delayed acks, but disabling those is a receive-side
-        // operation (TCP_QUICKACK), and we can't always control the
-        // client. PostgreSQL sets TCP_NODELAY on both sides of its
-        // sockets, so it seems sane to just do the same.
-        //
-        // If set_nodelay fails, it's a programming error, so panic.
-        //
-        // [0]: https://news.ycombinator.com/item?id=10608356
-        conn.set_nodelay(true).expect("set_nodelay failed");
-        // Enable TCP keepalives to avoid any idle connection timeouts that may
-        // be enforced by networking devices between us and the client. Idle SQL
-        // connections are expected--e.g., a `SUBSCRIBE` to a view containing
-        // critical alerts will ideally be producing no data most of the time.
-        if let Err(e) = SockRef::from(&conn).set_tcp_keepalive(&KEEPALIVE) {
-            error!("failed enabling keepalive: {e}");
-            continue;
+            // Actively cull completed tasks from the JoinSet so it does not grow unbounded. This
+            // method is cancel safe.
+            res = set.join_next(), if set.len() > 0 => {
+                if let Some(Err(e)) = res {
+                    debug!(
+                        "error joining connection in {}: {}",
+                        S::NAME,
+                        e.display_with_causes()
+                    );
+                }
+            }
+        }
+    }
+    if let Some(wait) = wait_timeout {
```

Comment:
What do you think of using a default of 1 second, if no other duration is provided? I don't think `environmentd` nor `clusterd` do clean shutdowns today, but it could be nice to start working in that direction?

Reply:
I think that would be a policy change for a component unrelated to balancerd and thus should get its own PR. It sounds reasonable though!

Reply:
Talked with Alex about this: we think it would be good to do at ~5s or so for envd and is safe because:

1. In the case of a newer envd that fences out this old one, the grace period won't be used because envd will halt itself.
2. In the case of no new envd (say, due to a node bump in k8s), a few seconds of grace period will let short connections finish their work.